### PR TITLE
wal: fix ReleaseLockTo

### PR DESCRIPTION
**Pull request by xiang90**
ReleaseLockTo should not release the lock on the WAL
segment that is right before the given index. When
restarting etcd, etcd needs to read from the WAL segment
that has a smaller index than the snapshot index.

The correct behavior is that ReleaseLockTo releases
the locks w is holding so that w only holds one lock
that has an index smaller than the given index.

### CODE REVIEW

File: wal/wal.go [hunk 1]
```diff
@@ -330,27 +330,33 @@ func (w *WAL) sync() error {
...
+
+	if smaller <= 0 {
+		return nil
+	}
+
+	for i := 0; i < smaller; i++ {
+		w.locks[i].Unlock()
+		w.locks[i].Destroy()
+	}
+	w.locks = w.locks[smaller-1:]
```
Comment (yichengq) [inline wal/wal.go]: should this be `[smaller:]` because we just want the file at the `smaller` index?

File: wal/wal.go [hunk 2]
```diff
@@ -330,27 +330,33 @@ func (w *WAL) sync() error {
 	return err
 }
 
-// ReleaseLockTo releases the locks w is holding, which
-// have index smaller or equal to the given index.
+// ReleaseLockTo releases the locks w is holding so that
+// w only holds one lock that has an index smaller than
+// the given index.
```
Comment (yichengq) [inline wal/wal.go]: don't follow this sentence easily. I thought it doesn't hold the lock with bigger index at the first read.

File: wal/wal.go (hunk 1 above)
Comment (xiang90) [inline wal/wal.go]: sure. then i need to make the previous < to <=

**Comment by xiang90** [comment]
@yichengq Fixed. PTAL.

File: wal/wal.go [hunk 3]
```diff
@@ -330,27 +330,34 @@ func (w *WAL) sync() error {
...
+
+	for i, l := range w.locks {
+		_, lockIndex, err := parseWalName(path.Base(l.Name()))
 		if err != nil {
 			return err
 		}
-		err = l.Destroy()
-		if err != nil {
-			return err
+		if lockIndex >= index {
```
Comment (yichengq) [inline wal/wal.go]: I think `lockIndex > index` is sufficient, but `>=` is certainly safer. Just wanna make sure I understand this correctly.

**Comment by yichengq** [comment]
lgtm

File: wal/wal.go (hunk 3 above)
Comment (xiang90) [inline wal/wal.go]: it is not. 1,2,3,4,5 -> release 5 will end up with only 5 if > , which is not what described in the comment.

**Pull request merged**